### ISSUE TITLE
InputNumber: Long/Integer values should implicitly set decimalPlaces="0"

### ISSUE THREAD

**Issue by tuerker** [primefaces/primefaces#5578]
The inputNumber component

 <p:inputNumber value="#{givenValue}"/>

should not require to EXPLICITLY set the attribute decimalPlaces="0" in case the given value is a Long/Integer/Short value.

## 1) Environment
- PrimeFaces version: 8.0 RC3

## 2) Expected behavior

The inputNumber component should restrict decimalPlaces to zero whenever the component is invoked with a Long/Integer value.

## 3) Actual behavior

Currently, the inputNumber component allows to enter numeric values with a fraction of 2 for Long/Integer values.
..

## 4) Steps to reproduce

.. 

## 5) Sample XHTML

..

## 6) Sample bean

..

**Comment by melloware** [comment]
This is a good idea I just ran into this myself.

**Comment by tuerker** [comment]
Ideally, the min and max values of the underlying number class are respected, too, i.e., that corresponding minValue and maxValue attribute values are set explicitly. Of course, all this must be done before applying potential bean validation constraints.

**Comment by tuerker** [comment]
Could you please extend this also for BigInteger. Thanks in advance.

     if (value instanceof Long || value instanceof Integer  || value instanceof Short **|| value instanceof BigInteger**) {
            defaultDecimalPlaces = "0";
}

**Comment by tuerker** [comment]
@melloware: I would like to come back to my other issue, p:inputNumber: Too restrictive for double/float values #5573, which was simply closed yesterday. I do not agree that a default decimalPlaces of 2 is a good solution per se. This might be the case for developers implementing software for the financial industry dealing with prices only but not for those working in the scientific field where decimalPlaces are usually "unbounded", i.e., the classes Double, Float, and BigDecimal (with no explicit restriction of the fraction part) are heavily used. For them, the p:inputNumber component cannot generically be used without doubling all the generic code fragments, depending whether the input values of the component are Integers (Z) or not. Actually, for Double, Float, and BigDecimal either one places an arbitrary big number of decimalPlaces on the p:inputNumber component or uses the p:inputText (with some regex) instead of the p:inputNumber component. In any case, an unsatisfactory solution.

However to avoid a long discussion. Since PrimeFaces relies on AutoNumeric.js, my question is: Does AutoNumeric.js really enforce to enter a value for decimalPlaces or can it be left open such that i can enter any decimal value? If AutoNumeric has this "financial legacy" of forcing defaultdecimalplaces=2 for any numbers, then there are two viable options in the long run:

1) PrimeFaces should think of replacing this library by a more powerful one, which can handle all types of numbers correctly.

or 

2) The AutoNumeric developers should generalize their library and fix this issue.

Many thanks for all your efforts.

**Comment by melloware** [comment]
To answer your questions...

- I closed yesterday's ticket because there is nothing for PrimeFaces to fix which I why i marked it WontFix.
- I updated for BigInteger
- Yes Autonumeric forces 2 decimalPlaces if you leave DecimalPlaces either NULL or Undefined.  There is no concept of unbounded decimal places.
- As @metaxmx said why not just set DecimalPlaces="25" or some large number if you need that.  Unbounded does not really make sense and I am not sure what type of application you have that you would need that small of a decimal number but there is nothing stopping you from setting a high number of decimal places
- AutoNumeric IS the most powerful and popular JS library for input number submission.  But feel free to look or suggest another one.  AutoNumeric has been around over 10 years and version 4 which we are using was a complete rewrite from the ground up
- feel free to submit an issue on the Autonumeric GithHub issues: https://github.com/autoNumeric/autoNumeric/issues

**Comment by tuerker** [comment]
Many thanks for your comments. Then, I will submit an issue on the Autonumeric GithHub.

**Comment by tuerker** [comment]
> As @metaxmx said why not just set DecimalPlaces="25" or some large number if you need that. 

This is exactly what we were doing and which is working only for individual cases.

However, since we are more and more reusing complex generic code fragments and do not want to copy the entire code fragments for all the different cases, it is essential that decimalPlaces MUST NOT be bounded by default.

Look at this code fragment:

```
<p:column
    id="#{columnId}#{constants.COLUMN}"
    rendered="#{rendered}"
    exportable="false"
    styleClass="#{columnId}#{rowIndex}">
    <f:facet name="header">
      <p:inputNumber
        id="#{columnId}#{constants.HEADER_INPUT}"
        disabled="#{columnInputEnabled != null and not columnInputEnabled or headerInputEnabled != null and not headerInputEnabled}"
        valueChangeListener="#{bean.valueChangedAll}"
        title="#{messages.setValueForColumnHint}"
        decimalPlaces="#{not empty decimalPlaces ? decimalPlaces : 32}"
        decimalSeparator="."
        thousandSeparator=""
        padControl="#{not empty padControl ? padControl : false}"
        onblur="this.value=''"
        inputStyleClass="table-header-margin"
        inputStyle="#{not empty style ? style : ''}">
        <p:ajax
          event="change"
          partialSubmit="true"
          resetValues="true"
          update="@this, #{columnId}#{constants.WATERMARK_FIELD}#{constants.HEADER}" />
      </p:inputNumber>
      <h:panelGroup id="#{columnId}#{constants.WATERMARK_FIELD}#{constants.HEADER}">
        <p:watermark
          for="#{columnId}#{constants.HEADER_INPUT}"
          value="#{(columnInputEnabled != null and not columnInputEnabled or headerInputEnabled != null and not headerInputEnabled) ? '' : messages.setValueForColumnLabel}" />
      </h:panelGroup>
      <ui:include src="/fragments/display-required.xhtml">
        <ui:param
          name="label"
          value="#{facetLabel}" />
        <ui:param
          name="required"
          value="#{displayRequired}" />
      </ui:include>
    </f:facet>
    <h:panelGroup id="#{columnId}#{constants.REQUIRED_FIELD}">
      <h:outputText
        rendered="#{not empty displayRequiredCell and displayRequiredCell}"
        value="&#160;*"
        title="#{messages.requiredHint}"
        styleClass="required" />
    </h:panelGroup>
    <p:inputNumber
      id="#{columnId}#{constants.INPUT}"
      disabled="#{columnInputEnabled != null and not columnInputEnabled or cellInputEnabled != null and not cellInputEnabled}"
      value="#{not empty hideValueIfFieldDisabled and hideValueIfFieldDisabled and (columnInputEnabled != null and not columnInputEnabled or cellInputEnabled != null and not cellInputEnabled) ? null : value}"
      valueChangeListener="#{bean.valueChanged}"
      decimalPlaces="#{not empty decimalPlaces ? decimalPlaces : 32}"
      decimalSeparator="."
      thousandSeparator=""
      padControl="#{not empty padControl ? padControl : false}"
      inputStyleClass="inline-input #{columnId}#{constants.INPUT}#{rowIndex}"
      inputStyle="text-align: right; #{not empty style ? style : ''}">
      <c:if test="#{attributeCouplingEnabled == null or not attributeCouplingEnabled}">
        <p:ajax
          event="change"
          global="false"
          ignoreAutoUpdate="true"
          partialSubmit="true" />
      </c:if>
      <c:if test="#{attributeCouplingEnabled != null and attributeCouplingEnabled}">
        <p:ajax
          event="change"
          global="false"
          ignoreAutoUpdate="true"
          partialSubmit="true"
          listener="#{bean.updateCoupling(entity, columnId)}" />
      </c:if>
    </p:inputNumber>
    <h:panelGroup id="#{columnId}#{constants.WATERMARK_FIELD}">
      <p:watermark
        rendered="#{not empty watermarkValue}"
        for="#{columnId}#{constants.INPUT}"
        value="#{watermarkValue}" />
    </h:panelGroup>
    <div>
      <h:message
        id="#{columnId}#{constants.MESSAGE}"
        for="#{columnId}#{constants.INPUT}"
        styleClass="errorMessageField" />
    </div>
  </p:column>
```

Here, decimalPlaces is set to 32 by default. But actually, we do not want to restrict the decimalPlaces since it is too much in some cases while in others it is too less.

When this case fragment is invoked on any integer value, then decimalPlaces should be zero by default, which is now the case after your fixes. However, when it is invoked on a float, double, or bigdemical (not restricted by @Digits) value, the decimalPlaces cannot be restricted by any artificial value. However, if the attribute decimalPlaces is null/empty, then it is two by default. So you have to conclude that the p:inputNumber component is not generically usable for float, double, and bigdecimal values unless you restrict them to an artificial size which then cannot be consistent with all your numeric Java/SQL attributes.
